Trigger toggle capture actions in case page muted state changes

youennf opened the pull request:
#### 6b9d0d0c4172149e7d2fc204a083ca16c27e3b16
<pre>
Trigger toggle capture actions in case page muted state changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=277261">https://bugs.webkit.org/show_bug.cgi?id=277261</a>
<a href="https://rdar.apple.com/132725520">rdar://132725520</a>

Reviewed by Jean-Yves Avenard.

When user is toggling capture mute/unmute buttons, this triggers a call to Page::setMuted.
We pipe this information to MediaSession.
We make sure to call the MediaSession action handlers:
- for documents that have live capture tracks
- before live capture tracks fire the corresponding mute/unmute event.

* LayoutTests/http/wpt/mediasession/resources/toggleCapture-iframe.html: Added.
* LayoutTests/http/wpt/mediasession/toggleCapture-expected.txt: Added.
* LayoutTests/http/wpt/mediasession/toggleCapture.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::platformCommandForMediaSessionAction):
(WebCore::MediaSession::callActionHandler):
* Source/WebCore/Modules/mediasession/MediaSessionAction.h:
* Source/WebCore/Modules/mediasession/MediaSessionAction.idl:
* Source/WebCore/Modules/mediasession/MediaSessionActionDetails.h:
* Source/WebCore/Modules/mediasession/MediaSessionActionDetails.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::hasRealtimeMediaSource):
(WebCore::Document::processCaptureStateDidChange):
(WebCore::Document::cameraCaptureStateDidChange):
(WebCore::Document::microphoneCaptureStateDidChange):
(WebCore::Document::screenshareCaptureStateDidChange):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMuted):

Canonical link: <a href="https://commits.webkit.org/281596@main">https://commits.webkit.org/281596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1409c7ae9ae1ccfaba3dff755c4c952b88acb7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39715 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12922 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64283 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11128 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48870 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7588 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37012 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52300 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29717 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33710 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9525 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9812 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66015 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9649 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56228 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4317 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52273 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56397 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13423 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3581 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35525 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37696 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on LayoutTests/http/wpt/mediasession/toggleCapture.html:
```diff
@@ -0,0 +1,146 @@
+
+async function testActionEventOrder(track, action)
+{
+    let isActionHandled = false;
```
can we move the test for `window.internals` here?

youennf replied:
Done

jyavenard commented on LayoutTests/http/wpt/mediasession/toggleCapture.html:
```diff
@@ -0,0 +1,146 @@
+}
+
+promise_test(async (test) => {
+    if (!window.internals)
```
would be nicer to not test for window.internals here, would make migration to a UA-agnostic test easier

youennf replied:
Done

jyavenard commented on LayoutTests/http/wpt/mediasession/toggleCapture.html:
```diff
@@ -0,0 +1,146 @@
+}, "togglecamera and mute/unmute events order");
+
+promise_test(async (test) => {
+    if (!window.internals)
```
same

jyavenard commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5323,6 +5326,73 @@ void Document::visibilityAdjustmentStateDidChange()
+        return;
+
+    eventLoop().queueTask(TaskSource::MediaElement, [weakDocument = WeakPtr { *this }, weakSession = WeakPtr { *mediaSession }, isPageMuted = isPageMutedCallback(*page), filterSource = WTFMove(filterSource), isPageMutedCallback = WTFMove(isPageMutedCallback), action] {
+        RefPtr protecteDocument = weakDocument.get();
```
checking for protectedDocument here would improve readability

youennf replied:
Done

jyavenard reviewed (approved):
LGTM

youennf committed:
Trigger toggle capture actions in case page muted state changes
https://bugs.webkit.org/show_bug.cgi?id=277261
rdar://132725520

Reviewed by Jean-Yves Avenard.

When user is toggling capture mute/unmute buttons, this triggers a call to Page::setMuted.
We pipe this information to MediaSession.
We make sure to call the MediaSession action handlers:
- for documents that have live capture tracks
- before live capture tracks fire the corresponding mute/unmute event.

* LayoutTests/http/wpt/mediasession/resources/toggleCapture-iframe.html: Added.
* LayoutTests/http/wpt/mediasession/toggleCapture-expected.txt: Added.
* LayoutTests/http/wpt/mediasession/toggleCapture.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::platformCommandForMediaSessionAction):
(WebCore::MediaSession::callActionHandler):
* Source/WebCore/Modules/mediasession/MediaSessionAction.h:
* Source/WebCore/Modules/mediasession/MediaSessionAction.idl:
* Source/WebCore/Modules/mediasession/MediaSessionActionDetails.h:
* Source/WebCore/Modules/mediasession/MediaSessionActionDetails.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::hasRealtimeMediaSource):
(WebCore::Document::processCaptureStateDidChange):
(WebCore::Document::cameraCaptureStateDidChange):
(WebCore::Document::microphoneCaptureStateDidChange):
(WebCore::Document::screenshareCaptureStateDidChange):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMuted):

Canonical link: https://commits.webkit.org/281596@main
